kernel/rros/poll.rs

```diff
@@ -0,0 +1,1201 @@
+    pub filp: Option<File>,
+    pub node: RrosPollNode,
+}
+
+impl RrosPollWatchpoint {
+    pub fn new() -> Self {
+        Self {
+            fd: 0,
+            events_polled: 0,
+            pollval: None,
+            wait: OobPollWait::new(),
+            flag: None,
+            filp: None,
+            node: RrosPollNode::new(),
+        }
+    }
+}
+pub struct RrosPollCtlreq {
+    pub action: u32,
+    pub fd: u32,
+    pub events: u32,
+    pub pollval: RrosValue,
+}
+
+impl RrosPollCtlreq {
+    pub fn new() -> Self {
+        Self {
+            action: 0,
+            fd: 0,
+            events: 0,
+            pollval: RrosValue::new(),
+        }
+    }
+}
+
+#[repr(C)]
+struct CPollEvent {
+    fd: u32,
+    events: u32,
+    pollval: Pollvalue,
+}
+
+impl CPollEvent {
+    fn new() -> Self {
+        Self {
+            fd: 0,
+            events: 0,
+            pollval: Pollvalue {
+                ptr: core::ptr::null_mut(),
+            },
+        }
+    }
+}
+#[repr(C)]
+union Pollvalue {
+    val: i32,
+    lval: i64,
+    ptr: *mut c_types::c_void,
+}
+
+pub struct RrosPollEvent {
+    pub fd: u32,
+    pub events: u32,
+    pub pollval: Option<Rc<RefCell<RrosValue>>>,
+}
+
+impl RrosPollEvent {
+    pub fn new() -> Self {
+        Self {
+            fd: 0,
+            events: 0,
+            pollval: None,
+        }
+    }
+}
+
+#[repr(C)]
+pub struct RrosPollWaitreq {
+    pub timeout_ptr: u64,
+    pub pollset_ptr: u64,
+    nrset: i32,
+}
+
+impl RrosPollWaitreq {
+    pub fn new() -> Self {
+        Self {
+            timeout_ptr: 0,
+            pollset_ptr: 0,
+            nrset: 0,
+        }
+    }
+}
+
+impl RrosThread {
+    pub fn drop_poll_table(&mut self) {
+        match self.poll_context.table.as_ref() {
+            Some(table) => {
+                drop(table);
+            }
+            None => (),
+        }
+    }
+}
+
+fn connect_watchpoint(
+    wait: &mut OobPollWait,
+    head: NonNull<RrosPollHead>,
+    unwatch: Option<fn(head: &RrosPollHead) -> ()>,
+) -> () {
+    let mut i = 0;
+    for poco in wait.connectors.as_mut_slice() {
+        if poco.head.is_none() {
+            poco.head = Some(head);
+            poco.unwatch = unwatch;
+            poco.events_received = 0;
+            list_add!(
+                &mut poco.next,
+                head.as_ref().watchpoints.locked_data().get()
+            );
+            poco.index = i as i32;
+            return;
+        }
+        if poco.head.unwrap().as_ptr() == head.as_ptr() {
+            pr_err!("poll: connect_watchpoint: duplicated head\n");
+            return;
+        }
+        i += 1;
+    }
+    pr_err!("poll: connect_watchpoint: no free connectors\n");
+}
+
+pub fn rros_poll_watch(
+    head: NonNull<RrosPollHead>,
+    wait: &mut OobPollWait,
+    unwatch: Option<fn(head: &RrosPollHead) -> ()>,
+) -> () {
+    let flags: u64;
+
+    unsafe {
+        flags = head.as_ref().watchpoints.irq_lock_noguard();
+    }
```

Comment:
Currently, the `unsafe` blocks are due to the usage of the `list_head`.

Reply:
The `unsafe` block in **line 364** is due to `NonNull`

Reply:
Maybe I can just use `&mut` instead of `NonNull`. I will take a look.

Reply:
The type of `head` field in struct `RrosPollConnector` is `Option<NonNull<RrosPollHead>>` , I need a `NonNull` value for the assignment